spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionTensorflowOutputConverter.java

```diff
@@ -0,0 +1,125 @@
+import org.tensorflow.Tensor;
+
+import org.springframework.cloud.stream.app.tensorflow.processor.TensorflowOutputConverter;
+import org.springframework.core.io.Resource;
+import org.springframework.tuple.Tuple;
+import org.springframework.tuple.TupleBuilder;
+import org.springframework.util.Assert;
+import org.springframework.util.StreamUtils;
+
+import com.google.common.collect.Ordering;
+import com.google.common.primitives.Floats;
+
+/**
+ * @author Christian Tzolov
+ */
+public class ImageRecognitionTensorflowOutputConverter implements TensorflowOutputConverter<Tuple> {
+
+	private static final Log logger = LogFactory.getLog(ImageRecognitionTensorflowOutputConverter.class);
+
+	private final List<String> labels;
+
+	private final int responseSize;
+
+	public ImageRecognitionTensorflowOutputConverter(Resource labels, int responseSize) {
+		this.responseSize = responseSize;
+		try (InputStream is = labels.getInputStream()) {
+			this.labels = Arrays.asList(StreamUtils.copyToString(is, Charset.forName("UTF-8")).split("\n"));
+			Assert.notNull(this.labels, "Failed to initialize the labels list");
+		}
+		catch (IOException e) {
+			throw new RuntimeException("Failed to initialize the Vocabulary", e);
+		}
+
+		logger.info("Word Vocabulary Initialized");
+	}
+
+	@Override
+	public Tuple convert(Tensor tensor, Map<String, Object> processorContext) {
+		final long[] rshape = tensor.shape();
+		if (tensor.numDimensions() != 2 || rshape[0] != 1) {
+			throw new RuntimeException(
+					String.format(
+							"Expected model to produce a [1 N] shaped tensor where N is the number of labels, " +
+									"instead it produced one with shape %s",
+							Arrays.toString(rshape)));
+		}
+		int labelsCount = (int) rshape[1];
+
+		float[] labelProbabilities = tensor.copyTo(new float[1][labelsCount])[0];
+
+		List<Tuple> tuples =new ArrayList<>(responseSize);
+		if (responseSize == 1) {
+			int maxProbabilityIndex = maxProbabilityIndex(labelProbabilities);
+			tuples.add(TupleBuilder.tuple().of(labels.get(maxProbabilityIndex), labelProbabilities[maxProbabilityIndex]));
+		}
+		else {
+			List<Integer> top10Probabilities = topKProbabilities(labelProbabilities, responseSize);
+
+			for (int i = 0; i < top10Probabilities.size(); i++) {
+				int probabilityIndex = top10Probabilities.get(i);
+				tuples.add(TupleBuilder.tuple().of(labels.get(probabilityIndex), labelProbabilities[probabilityIndex]));
+			}
+		}
+
+		return TupleBuilder.tuple().of("labels", tuples);
+
+	}
+
+	private List<Integer> topKProbabilities(final float[] labelProbabilities, int k) {
+
+		List<Integer> list = new ArrayList<>(labelProbabilities.length);
+		for (int i = 0; i < labelProbabilities.length; i++) {
+			list.add(i);
+		}
+
+		List<Integer> topK = new Ordering<Integer>() {
+			@Override
+			public int compare(Integer left, Integer right) {
+				return Floats.compare(labelProbabilities[left], labelProbabilities[right]);
+			}
+		}.greatestOf(list, k);
```

Comment:
Well, we definitely can avoid Guava for this algorithm.

Reply:
+1